File: find-routes.py
#!/usr/bin/python3

# find routes that hit a given stop ("Grand Central - 42 St")

import csv


def main():
    with open('stops.txt') as stops_fh:
        stop_ids = set(find_stop_ids(stops_fh, 'Grand Central - 42 St'))
    assert len(stop_ids) > 0, 'no stops found'
    print('stop_ids:', stop_ids)

    with open('stop_times.txt') as stoptimes_fh:
        trip_ids = set(find_trip_ids(stoptimes_fh, stop_ids))
    print('trip_ids:', len(trip_ids))

    with open('trips.txt') as trips_fh:
        for route_id in find_routes(trips_fh, trip_ids):
            print(route_id)


def find_stop_ids(stops_fh, want_name):
    reader = csv.reader(stops_fh)
    headers = next(reader)
    id_idx = headers.index('stop_id')
    name_idx = headers.index('stop_name')
    for line in reader:
        if line[name_idx] == want_name:
            yield line[id_idx]


def find_trip_ids(stoptimes_fh, stop_ids):
    reader = csv.reader(stoptimes_fh)
    headers = next(reader)
    id_idx = headers.index('stop_id')
    trip_id_idx = headers.index('trip_id')
    for line in reader:
        if line[id_idx] in stop_ids:
            yield line[trip_id_idx]


def find_routes(trips_fh, trip_ids):
    reader = csv.reader(trips_fh)
    headers = next(reader)
    trip_id_idx = headers.index('trip_id')
    route_id_idx = headers.index('route_id')
    seen = set()
    for line in reader:
        if line[trip_id_idx] in trip_ids:
            route_id = line[route_id_idx]
            if route_id not in seen:
                yield route_id
                seen.add(route_id)


main()

File: find-routes.js
#!/usr/local/bin/node

// find routes that hit a given (hardcoded) stop
//
// Uses synchronous filesystem and csv-parse APIs, because it's the
// easiest way to do it. But this version is 5x slower than the Python
// version. Unclear why: maybe the use of synchronous I/O is to blame?
// No evidence for that, but I have a hard time believing that reading
// files in Python is 5x faster than in Node.
//
// (The CSV parser here is doing a bit more work, and making them
// equivalent yielded roughly 20% speedup -- nowhere close to 5x! So
// that's not the whole story.)

const fs = require('fs');
const csvparse = require('csv-parse/lib/sync');

function main() {
  let stop_ids = find_stop_ids("stops.txt", "Grand Central");
  let trip_ids = find_trip_ids("stop_times.txt", stop_ids);
  let route_ids = find_routes("trips.txt", trip_ids);
  for (let route_id of route_ids) {
    console.log(route_id);
  }
}

// find stops whose name starts with prefix; return set of stop IDs
// (string)
function find_stop_ids(filename, prefix) {
  let records = read_csv(filename);
  let results = new Set();
  for (let record of records) {
    if (record.stop_name.startsWith(prefix)) {
      results.add(record.stop_id);
    }
  }
  return results;
}

// find trips that hit a stop in stop_ids (a set); return set of trip
// IDs (string)
function find_trip_ids(filename, stop_ids) {
  let records = read_csv(filename);
  let results = new Set();
  for (let record of records) {
    if (stop_ids.has(record.stop_id)) {
      results.add(record.trip_id);
    }
  }
  return results;
}

// find routes that include a trip in trip_ids; return set of route
// IDs (string)
function find_routes(filename, trip_ids) {
  let records = read_csv(filename);
  let results = new Set();
  for (let record of records) {
    if (trip_ids.has(record.trip_id)) {
      results.add(record.route_id);
    }
  }
  return results;
}

// open filename, read it completely into memory, and return an array
// of records (one object per line, excluding the header line)
function read_csv(filename) {
  let data = fs.readFileSync(filename, {encoding: "utf8"});
  return csvparse(data, {columns: true});
}

main();
